includes/discovery/services.inc.php

```diff
@@ -1,35 +1,54 @@
 <?php
+/**
+ * services.inc.php
+ *
+ * Creates the correct handler for the trap and then sends it the trap.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Autodiscovers TCP services on devices in the 'server' device type.
+ * To use set in config.php:
+ * [discover_services] = 'true';
+ *
+ * @package    LibreNMS
+ * @link       http://librenms.org
+ * @copyright  2019 KanREN, Inc.
+ * @author     Heath Barnhart <hbarnhart@kanren.net>
+ */
 
 use LibreNMS\Config;
 
 if (Config::get('discover_services')) {
-    // FIXME: use /etc/services?
-    $known_services = array(
-        22  => 'ssh',
-        25  => 'smtp',
-        53  => 'dns',
-        80  => 'http',
-        110 => 'pop',
-        143 => 'imap',
-    );
-
     // Services
     if ($device['type'] == 'server') {
-        $oids = trim(snmp_walk($device, '.1.3.6.1.2.1.6.13.1.1.0.0.0.0', '-Osqn'));
+        $oidV4 = trim(snmp_walk($device, '.1.3.6.1.2.1.6.20.1.4.1.4.0.0.0.0', '-Osqn'));
+        $oidV6 = trim(snmp_walk($device, '.1.3.6.1.2.1.6.20.1.4.2.16.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0', '-Osqn'));
+        $oids = $oidV4 . $oidV6;
         foreach (explode("\n", $oids) as $data) {
             $data = trim($data);
             if ($data) {
                 list($oid, $tcpstatus) = explode(' ', $data);
-                if (trim($tcpstatus) == 'listen') {
-                    $split_oid = explode('.', $oid);
-                    $tcp_port  = $split_oid[(count($split_oid) - 6)];
-                    if ($known_services[$tcp_port]) {
-                        discover_service($device, $known_services[$tcp_port]);
-                    }
+                $split_oid = explode('.', $oid);
+                $tcp_port  = $split_oid[(count($split_oid) - 1)];
+                if (getservbyport($tcp_port, 'tcp')) {
+                    $services[] = getservbyport($tcp_port, 'tcp');
```

Comment:
Why add it to an array and then loop the array instead of just discovering inside this loop?